### DISCUSSION TITLE
Accessing Global Variable in Trial

### DISCUSSION THREAD

**Discussion by isabelletelford** [jspsych/jsPsych#1473] [Q&A]
Hi there,

I have a global variable `correct_trials` that is updated at different parts of my experiment to determine how many of the headphone check trials (the headphone check is 6 audio-button-response trials to determine if the participant's audio is good enough to complete the experiment) in the past headphone check node (includes all 6 trials) were correct. It's then used to decide if they have passed or failed the headphone check, and to determine whether they repeat it (after which there are other possible paths the experiment can take depending on the success/failure of that second headphone check) or go on to the main experiment.

If they fail the check the first time, an instructions trial is run telling the participant how many trials of the 6 they got right. At this point, `correct_trials` is updating correctly (if I do a `console.log(correct_trials)` the correct number is printed, and it recognizes if you passed the headphone check, skipping over the instructions trial as expected), but when it is used in the pages part of the instructions trial, it is always displayed as whatever the variable is initialized as. This is what is currently not working: `You have failed the headphone check, getting ' + correct_trials + ' out of 6 trials correct. You need at least 5 correct to pass.` No matter how many trials were correct, if you fail the check (get <5 correct), it says 'You have failed the headphone check, getting 0 out of 6 trials correct' as `correct_trials` is currently initialized to be 0. I am not sure if I am misunderstanding the control flow of JavaScript/jsPsych, but I cannot figure out how to get this piece of text to include the most recent value of `correct_trials`, and would appreciate any insight into what's going on! Here is some of my code, I will try to include everything that has to do with this variable.

            //----------------------------------------------------------------------------------//
            // HEADPHONE CHECK STIMULI WITH METADATA
            //----------------------------------------------------------------------------------//
            var headphone_stimuli = [
                {stimulus: 'headphone_check/antiphase_HC_IOS.wav', correct_response: '2'},
                {stimulus: 'headphone_check/antiphase_HC_ISO.wav', correct_response: '1'},
                {stimulus: 'headphone_check/antiphase_HC_OIS.wav', correct_response: '2'},
                {stimulus: 'headphone_check/antiphase_HC_OSI.wav', correct_response: '1'},
                {stimulus: 'headphone_check/antiphase_HC_SIO.wav', correct_response: '0'},
                {stimulus: 'headphone_check/antiphase_HC_SOI.wav', correct_response: '0'}
            ];

            //----------------------------------------------------------------------------------//
            // HEADPHONE CHECK TRIAL
            // Adapted from https://github.com/mcdermottLab/HeadphoneCheck
            //----------------------------------------------------------------------------------//
           
            var do_hc = true;
            var correct_trials = 0;
            
            var headphone_check = {
                type: 'audio-button-response',
                stimulus: jsPsych.timelineVariable('stimulus'),
                prompt: 'Which of the three sounds was the quietest?',
                choices: ['1st','2nd','3rd'],
                response_ends_trial: true,
                post_trial_gap: 1000,
                data: {correct_response: jsPsych.timelineVariable('correct_response')},
                on_finish: function(data){
                    data.correct = (data.button_pressed === data.correct_response);
                }
            };

            var headphone_check_pass = {
                type: 'instructions',
                show_clickable_nav: true,
                pages: [
                    //P1
                    '<p>You passed the headphone check!</p>' +
                    '<p>Click <strong>Next</strong> to continue.</p>',
                    ],
                on_start: function() {
                    jsPsych.setProgressBar(0.08);
                },
            };

            //----------------------------------------------------------------------------------//
            // OPTIONS IF HEADPHONE CHECK FAILED A SECOND TIME
            //----------------------------------------------------------------------------------//
            
            // create variable to determine which node plays next:
            //  0 = redo headphone check
            //  1 = continue to experiment as if they had passed check; data will be marked as unusable
            //  2 = continue to debrief to get SONA credit
            var fail_choice;
            
            var fail_options = {
                type: 'html-button-response',
                stimulus: 'Options:',
                prompt: 'You have now failed the headphone check twice. You may either retry the headphone check; continue to the experiment, but know that your data will not be used by the experimenters (you will still get SONA credit); or be redirected to SONA where you will get full credit for attempting the experiment.',
                choices: ['Try headphone check again', 'Continue to experiment', 'Redirect to SONA'],
                on_start: function() {
                    fail_choice = null;
                },
                on_finish: function(trial_data) {
                    var response = trial_data.button_pressed
                    if (response == 0) {
                        do_hc = true;
                        fail_choice = 0;
                    } else if (response == 1) {
                        fail_choice = 1;
                    } else if (response == 2) {
                        fail_choice = 2;
                    }
                }
            };

            var fail_options_node = {
                timeline: [fail_options],
                conditional_function: function() {
                    correct_trials = jsPsych.data.get().last(6).filter({
                        correct: true,
                        trial_type: 'audio-button-response'
                    }).count();
                    return (correct_trials < 5);
                }
            };

            //----------------------------------------------------------------------------------//
            // HEADPHONE CHECK BLOCK
            //----------------------------------------------------------------------------------//

            var headphone_check_node = {
                timeline: [headphone_check],
                timeline_variables: headphone_stimuli,
                randomize_order: true,
                repetitions: 1,
                conditional_function: function() {
                    return (do_hc);
                }
            };

            var headphone_check_loop = {
                timeline: [headphone_check_node, fail_options_node],
                loop_function: function() {
                     correct_trials = jsPsych.data.get().last(6).filter({
                     correct: true,
                     trial_type: 'audio-button-response'
                 }).count();
                    return (correct_trials < 5);
                },
                conditional_function: function() {
                    return (fail_choice == 0)
                }

            }

            //----------------------------------------------------------------------------------//
            // HEADPHONE CHECK PROCEEDINGS **IF PASSED**
            //----------------------------------------------------------------------------------//
            var hc_pass = {
                timeline: [
                    headphone_check_pass, 
                    instructions, 
                    practice, 
                    trial_instructions, 
                    block, 
                   upload_data, 
                    debrief],
                conditional_function: function() {
                    correct_trials = jsPsych.data.get().last(6).filter({
                    correct: true,
                    trial_type: 'audio-button-response'
                }).count();
                    return (correct_trials >= 5);
                }
            };

            var hc_pass_failed_hc = {
                timeline: [
                    instructions, 
                    practice, 
                    trial_instructions, 
                    block, 
                   upload_failed_data, 
                    debrief],
                conditional_function: function() {
                    return (fail_choice == 1);
                }
            };

            //----------------------------------------------------------------------------------//
            // HEADPHONE CHECK PROCEEDINGS **IF DID NOT PASS**
            //----------------------------------------------------------------------------------//
            
            var hc_round_2 = {
                timeline: [headphone_check],
                timeline_variables: headphone_stimuli,
                randomize_order: true,
                repetitions: 1
            };

            var headphone_check_fail = {
                type: 'instructions',
                show_clickable_nav: true,
                pages: [
                    '<p>You have failed the headphone check, getting ' + correct_trials + ' out of 6 trials correct. You need at least 5 correct to pass.</p>' +
                    '<p>You will now repeat the headphone check. Make sure you are in a quiet location with headphones on. Click <strong>"Next"</strong> to continue.</p>'
                ],
                on_start: function() {
                    correct_trials = jsPsych.data.get().last(6).filter({
                        correct: true,
                        trial_type: 'audio-button-response'
                    }).count();
                },
                on_finish: function() {
                    console.log(correct_trials);
                }
            };

            
            var hc_fail = {
                timeline: [
                    headphone_check_fail,
                    hc_round_2,
                    hc_pass,
                    fail_options_node,
                    headphone_check_loop,
                    hc_pass_failed_hc,
                    debrief_node
                ],
                on_start: function() {
                    do_hc = false;
                },
                conditional_function: function() {
                    correct_trials = jsPsych.data.get().last(6).filter({
                    correct: true,
                    trial_type: 'audio-button-response'
                }).count();
                    return (correct_trials < 5);
                }
            };
`

**Comment by jodeleeuw** [comment]
Hi @isabelletelford,

This is a case where you need [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters). This [video](https://www.youtube.com/watch?v=8-j2aAZ_iOk) goes into more detail about the flow and why this is necessary.

**Reply by isabelletelford** [reply]
Thank you! Fixed it:)
